Contributor/runtime, first version (#105)
decentraland/documentation: content/contributor/runtime/globals.md

---
title: "Globals"
sidebartitle: "Globals"
url: "/contributor/runtime/globals"
weight: 2
---

The scene runtime ensures the existence of certain objects and functions inside the sandboxed Javascript environment, in addition to the standard utilities such as `Promise`, `Date` or `Math`.

These are adapted implementations of familiar browser and/or Node concepts.

## Globals

The runtime injects 7 definitions in the scene's global scope:

> 1. [`console`](#console): a subset of the Node/browser `console` object.

Reviewer: ```suggestion
1. [`console`](#console): a simple `console` object, implementing a subset of functions of the Browser/Node.js console.
```

> 2. [`module`](#module): a subset of Node's `module` object.

Reviewer: ```suggestion
2. [`module`](#module): an object used to keep track of the `.exports` object.
```

> 3. [`exports`](#module): a shortcut to `module.exports`, as in Node.

Reviewer: ```suggestion
3. [`exports`](#module): an object that the module may add its API to as it executes. Compatible with [CommonJS Modules 1.0](https://wiki.commonjs.org/wiki/Modules/1.0)
```

> 4. [`require`](#module): an implementation of Node's `require` for a curated list of modules.

Reviewer: ```suggestion
4. [`require`](#module): a constrained implementation of the [CommonJS' Require](https://wiki.commonjs.org/wiki/Modules/1.0) function, used to load the external modules to interact with the engine, communications systems, the blockchain, etc.
```

> 5. [`setImmediate`](#scheduling): an implementation of Node's `setImmediate`.

Reviewer: ```suggestion
5. [`setImmediate`](#scheduling): an implementation of Node's `setImmediate`. Schedules to run handler immediately after running tasks have been flushed from the JS's event loop.
```

All of these are defined as read-only properties, so they cannot be reassigned. Some will throw exceptions when used unless certain [permissions]({{< ref "/contributor/content/entity-types/scenes#permissions" >}}) are granted to the scene.

## Console {#console}

Scenes have access to a `console` object, much like the one provided by a browser or Node environment, though limited to the some of the methods you would normally find.

```ts
type Console = {
  log(...args: any): void
  info(...args: any): void
  debug(...args: any): void
  warning(...args: any): void
  error(...args: any): void
}
```

Just like their standard counterparts, each method takes variable arguments of any type and renders them as human-readable messages. For example, this is valid:

```js
console.log("The thing just appeared", { thing: "foo" }, [1, 2, 3])
```

The precise behavior of these methods depends on the provider, but the messages must be accessible to scene developers debugging their code.

{{< info >}}
In the Foundation's browser-based World Explorer, log messages appear in the developer tools panel.
{{< /info >}}

## Imports and Exports {#module}

> Scenes can import and export objects using a Node-like interface.

Reviewer: ```suggestion
Scenes can interact with other system components via using a CommonJS-like `require` or imports in a high level language like Typescript.
```

We can also discuss how `import {x} from `~system/Asd`` becomes `const {x} = require(~system/Asd)` thanks to Typescript and how it helps the devexp

The `require` function provides access to optional modules that scenes can leverage. Since the `require` function is injected by the runtime, so are the module definitions and their import paths.

```ts

> require(moduleName: string): { exports: Object }

Reviewer: The exports is _the_ return type in this case
```suggestion
require(moduleName: string): ExportedObject
```

```

> Scenes can (and must]]) export objects by setting keys in `module.exports`, or the `exports` shortcut.

Reviewer: Here also is a good place to mention high level import

See the [execution]({{< relref "execution" >}}) page for more information.

## Scheduling {#scheduling}

> When a scene needs to schedule callbacks with their execution controlled by the sandbox host, it can use the Node-like `setImmediate` global. These callbacks will queue up and be invoked at some implementation-dependent point in the future, usually within one tick of the game loop.

Reviewer: I'd avoid mentioning node or browsers as much as possible to prevent confusions. In the past I had argues with people not being able to execute node.js modules OOTB in scenes.

```suggestion
When a scene needs to schedule callbacks with their execution controlled by the sandbox host, it can use the `setImmediate` global. These callbacks will queue up and be invoked at some implementation-dependent point in the future, usually at the end of the [game loop tick](https://adr.decentraland.org/adr/ADR-148).
```

## Imports and Exports {#module}

Scenes can import and export objects using the traditional [CommonJS](https://wiki.commonjs.org/wiki/Modules/1.0) module interface.

```ts
exports: Object
require(moduleName: string): Object
```

The `require` function allows scenes to access runtime-provided modules (such as [EngineApi]({{< relref "modules/engine_api" >}}) or [RestrictedActions]({{< relref "modules/restricted_actions" >}})), and nothing else (it **does not** access NPM packages or modules by path).

Properties added to the `exports` object are the scene's public interface and will be exposed to the runtime. In fact, scenes _must_ expose at least one method to run properly (see [execution]({{< relref "execution" >}})).

{{< info >}}
Scenes written in languages such as TypeScript use the more modern `import` and `export` statements, which can be transpiled into CommonJS-compatible uses of `require` and `exports`.

> {{< /info >}}

Reviewer: <3
